rock/src/io_stream.mli

```diff
@@ -0,0 +1,17 @@
+module Input : sig
+  type 'a t
+
+  val create : (unit -> 'a option Lwt.t) -> 'a t
+  val singleton : 'a -> 'a t
+  val read : 'a t -> 'a option Lwt.t
+  val iter : ('a -> unit Lwt.t) -> 'a t -> unit Lwt.t
```

Comment:
When working with Httpaf bodies, we'll have places where we use `iter` to consume the input stream, but the handler there will be of the form,

```ocaml
Input.iter (fun item -> Httpaf.Body.write_string item; Lwt.return_unit) stream;;
```

I think it might make sense to add a second flavor of iter that works with `('a -> unit)`. Maybe we can match the naming scheme from lwt by having 
```ocaml
val iter : ('a -> unit) -> 'a t -> unit Lwt.t
val iter_s : ('a -> unit Lwt.t) -> 'a t -> unit Lwt.t
```

Reply:
> Input.iter (fun item -> Httpaf.Body.write_string item; Lwt.return_unit) stream;;

I think this is bound to allocate a huge amount of memory though. We need to return a promise that's determined when the buffer hits max size and needs to be flushed.

Reply:
> I think it might make sense to add a second flavor of iter that works with ('a -> unit). Maybe we can match the naming scheme from lwt by having

For now let's see if we can make do with `iter` that returns a `unit Lwt.t`. I suspect that the other `iter` will not be very useful in the end.

Reply:
> I think this is bound to allocate a huge amount of memory though. We need to return a promise that's determined when the buffer hits max size and needs to be flushed.

True. We don't read this config today, but we should respect the body buffer size limits set by the user in the server config handed to Rock. Working with promises when writing to httpaf's body should straightforward. (Reading is a different story though).